thrust/thrust/complex.h

```diff
@@ -150,898 +46,330 @@ struct complex_storage;
-   */
-  template <typename U>
-  __host__ __device__
-  complex<T>& operator/=(const complex<U>& z);
-
-  /*! Adds a scalar to this \p complex and assigns the result to this
-   *  \p complex.
-   *
-   *  \param z The \p complex to be added.
-   *
-   *  \tparam U is convertible to \c value_type.
-   */
-  template <typename U>
-  __host__ __device__
-  complex<T>& operator+=(const U& z);
-
-  /*! Subtracts a scalar from this \p complex and assigns the result to
-   *  this \p complex.
-   *
-   *  \param z The scalar to be subtracted.
-   *
-   *  \tparam U is convertible to \c value_type.
-   */
-  template <typename U>
-  __host__ __device__
-  complex<T>& operator-=(const U& z);
-
-  /*! Multiplies this \p complex by a scalar and assigns the result
-   *  to this \p complex.
-   *
-   *  \param z The scalar to be multiplied.
-   *
-   *  \tparam U is convertible to \c value_type.
-   */
-  template <typename U>
-  __host__ __device__
-  complex<T>& operator*=(const U& z);
-
-  /*! Divides this \p complex by a scalar and assigns the result to
-   *  this \p complex.
-   *
-   *  \param z The scalar to be divided.
-   *
-   *  \tparam U is convertible to \c value_type.
-   */
-  template <typename U>
-  __host__ __device__
-  complex<T>& operator/=(const U& z);
-
-
-
-  /* --- Getter functions ---
-   * The volatile ones are there to help for example
-   * with certain reductions optimizations
-   */
-
-  /*! Returns the real part of this \p complex.
-   */
-  __host__ __device__
-  T real() const volatile { return data.x; }
-
-  /*! Returns the imaginary part of this \p complex.
-   */
-  __host__ __device__
-  T imag() const volatile { return data.y; }
-
-  /*! Returns the real part of this \p complex.
-   */
-  __host__ __device__
-  T real() const { return data.x; }
-
-  /*! Returns the imaginary part of this \p complex.
-   */
-  __host__ __device__
-  T imag() const { return data.y; }
-
-
-
-  /* --- Setter functions ---
-   * The volatile ones are there to help for example
-   * with certain reductions optimizations
-   */
+  __host__ __device__ complex(const base& rhs) : base(rhs) {}
+  __host__ __device__ complex(base&& rhs) : base(::cuda::std::move(rhs)) {}
 
-  /*! Sets the real part of this \p complex.
-   *
-   *  \param re The new real part of this \p complex.
-   */
+  template <class U>
   __host__ __device__
-  void real(T re) volatile { data.x = re; }
-
-  /*! Sets the imaginary part of this \p complex.
-   *
-   *  \param im The new imaginary part of this \p complex.e
-   */
-  __host__ __device__
-  void imag(T im) volatile { data.y = im; }
+    typename detail::enable_if<detail::and_<detail::has_promoted_numerical_type<T, U>,
+                                            detail::not_<detail::is_same<T, U>>>::value,
+                               complex>::type
+   &operator=(const complex<U> &rhs)
+  {
+    this->real(static_cast<T>(rhs.real()));
+    this->imag(static_cast<T>(rhs.imag()));
+    return *this;
+  }
```

Comment:
Nitpick: This requires instantiation of two types. We could improve compile time by moving the enable if into the template arguments:
```suggestion
  template <class U, typename detail::enable_if<!detail::is_same<T, U>::value, int> = 0,
                     typename detail::enable_if<detail::has_promoted_numerical_type<T, U>, int> = 0>
  __host__ __device__
  complex &operator=(const complex<U> &rhs)
  {
    this->real(static_cast<T>(rhs.real()));
    this->imag(static_cast<T>(rhs.imag()));
    return *this;
  }
```

This will short circuit if the first condition is not satisfied